### ISSUE TITLE
Why using restyle e4e latent instead of hyperstyle latent when running domain adaptation?

### ISSUE THREAD

**Issue by Shysto** [yuval-alaluf/hyperstyle#29]
Hi, first of all, thank you very much for this awesome project!

I have a question about the domain adaptation task.
If my understanding is correct, you [load](https://github.com/yuval-alaluf/hyperstyle/blob/77b25f64fadcc5e1a494cc36b665531d3763b06d/scripts/run_domain_adaptation.py#L31) the nets from restyle e4e in the `scripts/run_domain_adaptation.py` script.
This model is then used to [get the latent image](https://github.com/yuval-alaluf/hyperstyle/blob/77b25f64fadcc5e1a494cc36b665531d3763b06d/utils/domain_adaptation_utils.py#L10), from which you will generate the final image after domain adaptation.

My question is the following: after running some tests, the latent from Hyperstyle seems to be more faithful to the original image than the latent from restyle e4e, so why do you use the latent from restyle and [not the Hyperstyle one](https://github.com/yuval-alaluf/hyperstyle/blob/77b25f64fadcc5e1a494cc36b665531d3763b06d/utils/domain_adaptation_utils.py#L11)? 
Is there a reason to prefer the latent from restyle e4e for the domain adaptation task? 

Thank you for your answer.

**Comment by yuval-alaluf** [comment]
By HyperStyle's latent, I assume you mean the latent that you get from the WEncoder?

**Comment by Shysto** [comment]
@yuval-alaluf by HyperStyle's latent, I mean the second object that is returned by your call to the [`run_inversion()`](https://github.com/yuval-alaluf/hyperstyle/blob/77b25f64fadcc5e1a494cc36b665531d3763b06d/utils/domain_adaptation_utils.py#L11) function.  

If I am not mistaken, this function returns 4 values, the fourth (`codes`) corresponding to the latent from WEncoder that is used to [initialize](https://github.com/yuval-alaluf/hyperstyle/blob/77b25f64fadcc5e1a494cc36b665531d3763b06d/models/hyperstyle.py#L71) the latent object in the `forward()` function of your Hyperstyle net; while the second value (`result_latent`) is the [latent generated by Hyperstyle](https://github.com/yuval-alaluf/hyperstyle/blob/77b25f64fadcc5e1a494cc36b665531d3763b06d/models/hyperstyle.py#L85) after the weights of the generator have been updated by the hypernetwork. 
So I mean the latter, the latent that is retrieved by the inversion task (the latent representation of the image given by your `inference` script).

**Comment by yuval-alaluf** [comment]
The tuple here may be a bit confusing: 
```
y_hat, latent, weights_deltas, codes
```
In the end, `latent` and `codes` is the same thing. HyperStyle doesn't alter the latent code itself and only alters the weights of the generator. Therefore, HyperStyle's latents are just the WEncoder latents. That is, HyperStyle uses the same latent as the WEncoder in order to achieve its final reconstructions.

Therefore, to answer your question regarding why we chose to use the ReStyle-e4e latents instead of the WEncoder latents, it is simply because we found that it leads to slightly more accurate transformations. This is because ReStyle-e4e achieves better reconstructions compared to the vanilla WEncoder.
